src/PostgREST/Query/SqlFragment.hs

```diff
@@ -269,26 +269,33 @@ pgFmtCoerceNamed CoercibleField{cfName=fn} = pgFmtIdent fn
 
 pgFmtSelectItem :: QualifiedIdentifier -> CoercibleSelectField -> SQL.Snippet
 pgFmtSelectItem table CoercibleSelectField{csField=fld, csAggFunction=agg, csAggCast=aggCast, csCast=cast, csAlias=alias} =
-  pgFmtApplyAggregate agg aggCast (pgFmtApplyCast cast (pgFmtTableCoerce table fld)) <> pgFmtAs alias
+  pgFmtApplyAggregate agg aggCast Nothing (pgFmtApplyCast cast (pgFmtTableCoerce table fld)) <> pgFmtAs alias
 
-pgFmtSpreadSelectItem :: Alias -> SpreadSelectField -> SQL.Snippet
-pgFmtSpreadSelectItem aggAlias SpreadSelectField{ssSelName, ssSelAggFunction, ssSelAggCast, ssSelAlias} =
-  pgFmtApplyAggregate ssSelAggFunction ssSelAggCast fullSelName <> pgFmtAs ssSelAlias
+pgFmtSpreadSelectItem :: Alias -> MediaHandler -> SpreadSelectField -> SQL.Snippet
+pgFmtSpreadSelectItem aggAlias handler SpreadSelectField{ssSelName, ssSelAggFunction, ssSelAggCast, ssSelAlias} =
+  pgFmtApplyAggregate ssSelAggFunction ssSelAggCast (Just handler) fullSelName <> pgFmtAs ssSelAlias
   where
     fullSelName = case ssSelName of
       "*" -> pgFmtIdent aggAlias <> ".*"
       _   -> pgFmtIdent aggAlias <> "." <> pgFmtIdent ssSelName
 
-pgFmtApplyAggregate :: Maybe AggregateFunction -> Maybe Cast -> SQL.Snippet -> SQL.Snippet
-pgFmtApplyAggregate Nothing _ snippet = snippet
-pgFmtApplyAggregate (Just agg) aggCast snippet =
+pgFmtApplyAggregate :: Maybe AggregateFunction -> Maybe Cast -> Maybe MediaHandler -> SQL.Snippet -> SQL.Snippet
+pgFmtApplyAggregate Nothing _ _ snippet = snippet
+pgFmtApplyAggregate (Just agg) aggCast handler snippet =
   pgFmtApplyCast aggCast aggregatedSnippet
   where
     convertAggFunction = SQL.sql . BS.map toUpper . BS.pack . show
+    arrayAggStripNulls = case handler of
+      Just BuiltinAggArrayJsonStrip     -> True
+      Just (BuiltinAggSingleJson strip) -> strip
+      _                                 -> False
+    fmtArrayAggFunction
+      | arrayAggStripNulls = "array_agg(" <> snippet <> ") FILTER (WHERE " <> snippet <> " IS NOT NULL)"
+      -- TODO: NULLIF(...,'{null}') does not take into consideration a case with a single element with a null value.
+      -- See https://github.com/PostgREST/postgrest/pull/3640#issuecomment-2334996466
```

Comment:
Is this still a TODO? Is it not solved with the strip nulls?

Reply:
Yes. In both cases, when `nulls=stripped` and otherwise, the `[null]` will be converted to `[]`. For `nulls=stripped` this is expected, but when it's not used, then we would expect a single `[null]` if there's exactly one `null` value.

Reply:
Hm, is there someway to fix that here? Can we do something like `CASE array_length == 1 then [null]`?

Reply:
Hmm... `array_length` didn't work for me (it returned 1 for both cases), but using the `count` of the subquery seems to work. Just tested it with the example that I mentioned here https://github.com/PostgREST/postgrest/pull/3640#discussion_r1786835155.

The query would look something like:

```sql
SELECT "api"."factories"."name",

       -- Use count here instead of NULLIF(...)
       CASE WHEN "factories_processes_1".count = 0 THEN '{}' ELSE array_agg("factories_processes_1"."process_ids") END AS "process_ids",
       CASE WHEN "factories_processes_1".count = 0 THEN '{}' ELSE array_agg("factories_processes_1"."process_names") END AS "process_names"

FROM "api"."factories"
LEFT JOIN LATERAL
  (SELECT "processes_1"."id" AS "process_ids",
          "processes_1"."name" AS "process_names"
   FROM "api"."processes" AS "processes_1"
   WHERE "processes_1"."factory_id" = "api"."factories"."id") AS "factories_processes_1" ON TRUE
GROUP BY "api"."factories"."name";
```

It returns empty arrays when there are no elements, as expected. And when I `INSERT` the row I mentioned in my example, it returns `[null]` as expected. I'll check if I'm not missing something, otherwise this should be the fix.

Reply:
Hmm... the drawback I'm seeing with this is the cost increasing again due to the `COUNT`.

Reply:
I don't know anything about the code, but I know `COUNT` is not a viable option for large datasets, unless you use something like `LIMIT 1` with it?

Reply:
@laurenceisla How about using `json_agg_strict` introduced on pg 16? I've tried it and it works fine for removing the `[null]`. 

```sql
WITH pgrst_source AS (
    SELECT
      "test"."factories"."name",
      COALESCE(json_agg_strict("factories_processes_1"."process_names"),'{}') AS "process_names"
-- ...
```

We can document that `[null]` behavior as bug/gotcha on pg < 16.

Reply:
Problem about `json_agg_strict`, it removes all nulls even where cases the array length > 1. Sometimes nulls are needed to ensure order of the results as mentioned on https://github.com/PostgREST/postgrest/pull/3640#discussion_r1786943084

Reply:
@laurenceisla The COUNT is too expensive as Jonathan mentioned, another option might be doing the FILTER by PK as you mentioned before https://stackoverflow.com/questions/24155190/postgresql-left-join-json-agg-ignore-remove-null/33153530#33153530?

But if there's no PK or UNIQUE constraint, or if the user doesn't have the privilege to query those, then we would devolve again into the same case.

Reply:
> But if there's no PK or UNIQUE constraint, or if the user doesn't have the privilege to query those, then we would devolve again into the same case.

On second thought, the above should be a non-issue. To do the embed the FK has to exist and the user has to have the privileges otherwise the embed would fail anyway.

So the query should be like:

```sql
WITH pgrst_source AS (
    SELECT
      "test"."factories"."name",
      COALESCE(array_agg("factories_processes_1"."id") FILTER (WHERE factories_processes_1.factory_id IS NOT NULL),'{}') AS "id",
      COALESCE(array_agg("factories_processes_1"."process_names") FILTER (WHERE factories_processes_1.factory_id IS NOT NULL),'{}') AS "process_names"
    FROM "test"."factories"
    LEFT JOIN LATERAL (
        SELECT
            "processes_1"."id" AS "id",
            "processes_1"."name" AS "process_names",
            "processes_1"."factory_id"
        FROM "test"."processes" AS "processes_1"
        WHERE "processes_1"."factory_id" = "test"."factories"."id"
    ) AS "factories_processes_1" ON TRUE
    GROUP BY "test"."factories"."name"
)
SELECT
    null::bigint AS total_result_set,
    pg_catalog.count(_postgrest_t) AS page_total,
    coalesce(json_agg(_postgrest_t), '[]') AS body,
    nullif(current_setting('response.headers', true), '') AS response_headers,
    nullif(current_setting('response.status', true), '') AS response_status, '' AS response_inserted
FROM ( SELECT * FROM pgrst_source ) _postgrest_t;
```

A bit of a problem is that the sublevel must have always the FK in the `SELECT` and then it has to be removed in the upper level if the user didn't specify it; but that should be fixable. IIRC we already do the same for other features.